Use argument types as parameter types for inline closures (#7798)

leongersen opened the pull request:
Implements [#7798](https://github.com/phpstan/phpstan/issues/7798). Thank you for your advise, it was very helpful.

> The relevant code for this in NodeScopeResolver is this but you won't have to touch that for this feature request, it works in a general sense: https://github.com/phpstan/phpstan-src/blob/4b6d9647784966438c742b367066656cf269a8d4/src/Analyser/NodeScopeResolver.php#L2964-L2975

This unfortunately is not used for this case; `$passedToType` is only passed to `processClosureNode` when handling a closure as a function argument:
https://github.com/phpstan/phpstan-src/blob/4b6d9647784966438c742b367066656cf269a8d4/src/Analyser/NodeScopeResolver.php#L3223-L3226

https://github.com/phpstan/phpstan-src/blob/4b6d9647784966438c742b367066656cf269a8d4/src/Analyser/NodeScopeResolver.php#L2220-L2221

Because the closure is never the argument (it is the `name` of the `FuncCall`), this case is slightly different than for `array_map`, as the `Expr\Closure` is never handled by `ParametersAcceptorSelector::selectFromArgs` (only its `ClosureType` is).

Please let me know if I misunderstood.

leongersen committed:
Use argument types as parameter types for inline closures (#7798)

ondrejmirtes reviewed (changes requested):
This is really nicce and almost perfect! The only thing missing is that you should also test and make sure it works for arrow functions :) Please add another gathered file in NodeScopeResolverTest, but only for `PHP_VERSION_ID >= 70400`.

leongersen commented:
Thanks for the feedback, much appreciated. I've now added support and tests for arrow functions as well.

I've duplicated the logic and added another NodeVisitor, as `processArrowFunctionNode` duplicates logic from `processClosureNode`. If you'd like it abstracted I can of course do that.

ondrejmirtes commented:
Thank you!
